chain/client/src/client.rs

```diff
@@ -229,6 +242,91 @@ pub struct ProduceChunkResult {
     pub transactions_storage_proof: Option<PartialState>,
 }
 
+/// This keeps track of the numbef of new chunks seen in each shard since the block that was passed to new()
+/// This whole thing could be replaced with a much simpler function that just computes the numbef of new chunks
+/// in each shard from scratch every time we call it, but in the unlikely and unfortunate case where a shard
+/// hasn't had any chunks for a very long time, it would end up being a nontrivial inefficiency to do that
+/// every time run_catchup() is called
```

Comment:
So now that I think about your suggestion again, I think it would not quite result in the same sync hash (if I understnad your suggestion correctly, we just check `height_included` for each chunk and see if it's new or not? The issue I think is that we actually want two blocks with new chunks in every shard after the first block of the epoch (so usually the third block of the epoch). So unless I'm missing something, I think there has to be some sort of iteration. I think in another PR that fixes the FIXME in this one, we can figure something better out to reduce the number of places where this calculation is done (maybe for example saving this info for each block in the DB so we can just look it up, and when we apply a new block, we look up the info for its parent?). But if you have other ideas, I'm open to them